chain/sync.go

```diff
@@ -520,13 +520,31 @@ func (syncer *Syncer) Sync(ctx context.Context, maybeHead *types.TipSet) error {
 
 	hts := syncer.store.GetHeaviestTipSet()
 
+	// noop pre-checks
 	if hts.ParentWeight().GreaterThan(maybeHead.ParentWeight()) {
 		return nil
 	}
 	if syncer.Genesis.Equals(maybeHead) || hts.Equals(maybeHead) {
 		return nil
 	}
 
+	if maybeHead.Height() == hts.Height() {
+		// check if maybeHead is fully contained in headTipSet
+		// if that is the case, there is nothing for us to do
+		// we need to exit out early, otherwise checkpoint-fork logic will reject it
```

Comment:
I think this logic is right, but I'm not sure about the comment. The checkpoint issue is when `hts` is fully contained in `maybeHead`, not the other way around.

Here the issue is that we'll try to sync the wrong chain from our peers (when we later try to expand it?) and then _fail_, refusing to checkpoint to the smaller head. At least I think that's the issue...

Also, what happens if `maybeHead` is `hts+1`? `+2`?

Reply:
If our head is [A, B], and tipset [A] gets in here, it should be accepted, but the new checkpoint rejection logic will reject it.
If it is something built on [A, B] it is okay, checked in a test.

Reply:
It should be rejected because it's lighter than `[A, B]`. But this has nothing to do with checkpoints.

The _checkpointing_ issue is that, if we try to switch from `[A, B]` to `[A]`, we can fail in `Sync` in some test scenarios (that was the test issue _I_ was running into, at least) even though we already _have_ `[A]`. That will _prevent_ us from checkpointing.

So, I'm saying that this logic will fix the test issue but I'm still pretty sure the comment is wrong (and I'm not sure if this is the _best_ way to fix the failure).

Reply:
> It should be rejected because it's lighter than [A, B]. But this has nothing to do with checkpoints.

It will be accepted as we only look at ParentWeight. This is how tipset expansion naturally works. Without checkpoint given [A, B], incoming [C], expansion to [A, B, C] is natural.

Yes, there is another issue where if we are on [A, B] and we want to checkpoint [A], we will initiate fetch of [A]. This is more of a wart than a consensus-affecting bug.

1. Without this early escape, given head=checkpoint=[A, B], we would reject the sync attempt to [A], which should be accepted as it is a valid block that is part of our tipset (rejecting sync attempt of A might lead to marking [A] as bad block which we for sure don't want).
2. I've adjusted the test to push tipset [A,B] to the node and then checkpoint [A] to show that it works.
3. What this code fixes is rejecting the expansion from checkpointed [A] to [A, B], which is tested in [here](https://github.com/filecoin-project/lotus/pull/12747/files#diff-1444b0b32efb1403c113d6340498ebd7a2940b944011d25909c1871fb3922b04R1070)

Reply:
> The checkpoint issue is when hts is fully contained in maybeHead, not the other way around.

Yes, and that is resolved by flat-out rejecting tipsets that try to merge at or below checkpoint, but then, as explained in 1., we need to allow/or just ignore subsets of `hts`  while `hts = checkpoint` is our head, so we don't reject them.

Reply:
I slightly adjusted the wording of the comment

Reply:
Oh, I see... Can we fix this where we actually mark the tipset as bad? I.e.:

1. We check to see if it syncs away from the checkpoint.
2. Mark all blocks as bad that _aren't_ a part of the checkpoint?

Or am I still misunderstanding this issue.

Reply:
Blocks not part of the checkpoint will still get marked as bad. It is just the blocks that are part of the `checkpoint = head` that bypass sync.

Reply:
Also, I _think_ the current fix has a race:

1. We start syncing `[A]`.
2. We checkpoint `[A, B]`.
3. We finish syncing `[A]` and mark `[A]` as bad?

Reply:
They don't get added as bad. I would prefer not to touch that logic in this PR, though.
The `collectChain` errors don't get marked as bad automatically.